### PR TITLE
#171380748 Login with google and Facebook

### PR DESCRIPTION
**Pull request by Muhire-Josue**
#### What does this PR do?

- It allows the users to login using their google and Facebook accounts

#### Description of Task to be completed?

- It allows the user to login using their Facebook and Google accounts

#### How should this be manually tested?

- `git clone https://github.com/Stackup-Rwanda/stackup2-barefoot-backend.git`

- `cd stackup2-barefoot-backend`

- `git fetch origin ft-login-with-fb-and-google-171380748`

- `git checkout ft-login-with-fb-and-google-171380748`

- `npm install`

-  create a project on [developers facebook](https://developers.facebook.com/) and [developers google](https://developers.google.com/identity/sign-in/web/sign-in) in order to get the ID and Secret keys 

- Use `http://localhost:3000/api/auth/facebook/callback` as the callback URL for Facebook

- Use `http://localhost:3000/api/auth/google/callback` as the callback URL for Google

- create .env file and use .env.example file for reference of the environment variables

- npm run dev

- to test google auth use this URL in your browser: http://localhost:3000/api/auth/google/

- to test Facebook auth use this URL in your browser: http://localhost:3000/api/auth/facebook/
The result should be an object of the logged-in user

#### Any background context you want to provide?

- Please remember to create a project on [developers facebook](https://developers.facebook.com/) and [developers google](https://developers.google.com/identity/sign-in/web/sign-in) in order to get the ID and Secret keys, because these keys are the ones you will use to fill in the environment variables that are in .env.example file.
#### What are the relevant pivotal tracker stories?

- [171380748](https://www.pivotaltracker.com/story/show/171380748)

#### Screenshots (if appropriate)
facebook
google-1
google-2

### CODE REVIEW

File: src/controllers/auth.controller.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+/* eslint-disable require-jsdoc */
```
Comment (since9629) [inline src/controllers/auth.controller.js]: @Muhire-Josue  It is required that you use jsDoc to comment your code

File: package.json [hunk 2]
```diff
@@ -18,6 +18,7 @@
     "@babel/node": "^7.8.4",
     "@babel/preset-env": "^7.8.6",
     "@babel/register": "^7.8.6",
+    "babel-eslint": "^10.1.0",
```
Comment (since9629) [inline package.json:34]: @Muhire-Josue Why did you add this package?

Comment (Muhire-Josue) [reply package.json:34]: I used it to solve some warnings eslint was throwing about creating asyn/await functions

**Review by Ugizwenayo-Divine — COMMENTED** [review]
Josue, you did a pretty good job and difficult one but I suggested something
- you move those codes within src/controllers/auth.controller into  src/controllers/auth.controller as they are all for authentication purpose
- Also, move your routes within routes/api/authentication.routes.js file
- Finally the findUser function within src/utils, I suggest to be renamed identifyUser or something related to identifying because I've seen it's about checking what the user is going to login with.
And try to fix those coveralls issues

But actually you did an impressive work

File: public/api-docs/swagger.json [hunk 3]
```diff
@@ -107,83 +107,39 @@
         }
       }
     },
-    "/auth/resetpassword":{
```
Comment (mifeille) [inline public/api-docs/swagger.json:110]: Please check if you did not delete this route while rebasing

File: tests/socialLogin.test.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+);
+
+const facebookObject = authMockData[0];
+const googleObject = authMockData[1];
+
+describe('Passport Authentication controller', () => {
+  it('should authenticate a facebook user', (done) => {
+    chai
+      .request(app)
+      .post('/api/auth/facebook')
```
Comment (mifeille) [inline tests/socialLogin.test.js]: are you using the same route for testing? you should create test routes for testing

Comment (Muhire-Josue) [reply tests/socialLogin.test.js]: Am not using the same routes, this one is a Post but in the controller am using a Get. so different routes for tests and controller

File: public/api-docs/swagger.json (hunk 3 above)
Comment (Muhire-Josue) [reply public/api-docs/swagger.json:110]: I think I did, Let me bring it back

**Review by Ugizwenayo-Divine — COMMENTED** [review]
There some data you deleted from swagger.json file
and also solve those coeverall issues

File: package.json [hunk 5]
```diff
@@ -10,7 +10,9 @@
     "undo": "npx sequelize-cli db:migrate:undo",
     "seed": "npx sequelize-cli db:seed:all",
     "undo-seed": "npx sequelize-cli db:seed:undo",
-    "test": "npm run migrations && npm run seed && nyc mocha --require @babel/register ./tests/*.js  --timeout 25000 --exit && npm run undo-seed && npm run undo",
+    "pretest": "npm run migrations && npm run seed",
+    "test": "nyc mocha --require @babel/register ./tests/**/*.js  --timeout 30000 --exit",
```
Comment (descholar-ceo) [inline package.json]: I wonder why you changed these scripts of test, and how to run the test after you change them. You should document this change so that the team knows a new way of testing

File: public/api-docs/swagger.json [hunk 6]
```diff
@@ -280,74 +313,75 @@
...
+        "message": {
+          "type": "string",
+          "example": "enter your email or password"
         }
       }
-    }}
-}
+    }
+  }
+}
```
Comment (descholar-ceo) [inline public/api-docs/swagger.json]: Could you fix this newline issue?

File: src/database/migrations/20200309171605-create-user.js [hunk 7]
```diff
@@ -13,12 +13,21 @@ module.exports = {
     lastName: {
       type: Sequelize.STRING
     },
+    userName: {
```
Comment (descholar-ceo) [inline src/database/migrations/20200309171605-create-user.js]: Since it is no longer `userName` you should write `username` instead

File: src/database/migrations/20200309171605-create-user.js [hunk 8]
```diff
@@ -13,12 +13,21 @@ module.exports = {
...
+      type: Sequelize.STRING
+    },
     email: {
       type: Sequelize.STRING
     },
     password: {
       type: Sequelize.STRING
     },
+    social_media_id: {
```
Comment (descholar-ceo) [inline src/database/migrations/20200309171605-create-user.js]: Could you use camelCase for this `social media id`? For naming consistency.

File: src/database/models/user.js [hunk 9]
```diff
@@ -1,10 +1,12 @@
-
 module.exports = (sequelize, DataTypes) => {
   const user = sequelize.define('user', {
     firstName: DataTypes.STRING,
     lastName: DataTypes.STRING,
+    userName: DataTypes.STRING,
     email: DataTypes.STRING,
     password: DataTypes.STRING,
+    social_media_id: DataTypes.STRING,
```
Comment (descholar-ceo) [inline src/database/models/user.js]: Same here, you should use camelCase

File: src/routes/index.js [hunk 10]
```diff
@@ -1,8 +1,10 @@
 import express from 'express';
 import apiRouter from './api/index';
+// import passportAuthRoutes from './userAuth.router';
```
Comment (descholar-ceo) [inline src/routes/index.js:3]: Could you remove these commented line of codes?

File: src/routes/index.js [hunk 11]
```diff
@@ -1,8 +1,10 @@
...
 
 const allRoutes = express.Router();
 
 allRoutes.use('/api', apiRouter);
+// allRoutes.use(passportAuthRoutes);
```
Comment (descholar-ceo) [inline src/routes/index.js:8]: Same here, you should remove the commented lines of codes

File: src/utils/identifyUser.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
+import models from '../database/models';
+
+/**
+   * function findAll() returns all users in db
```
Comment (descholar-ceo) [inline src/utils/identifyUser.js]: is it function findAll() ?

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
@Muhire-Josue I personally like the job you did for this task, I have a few suggestions for you to take care of.
### 1. Changes on the PR structure
===>  In your PR in __how this PR should manually tested__ check again on the command you said to fetch a branch you said `git fetch ft-login-with-fb-and-google-171380748` but actually this command didn't work for me, I used `git fetch origin ft-login-with-fb-and-google-171380748`, if the command you put on the PR works for you, there are some other systems in which it doesn't work at all like mine, you should specify another alternative.
===>  This PR doesn't have enough details how I should run test locally, look at what I got after running `npm test`
Screenshot from 2020-03-13 00-45-34
I suggest you to add some steps to do before running `npm run dev` because since the tests aren't running because of my default database setup may be which is not adapted to your work, put some other details like using or changing a db name, because I finally found out that I was getting that error from my default db-name, so I had to use a different db and it worked, think of others who won't think about using different db, they will tend to think that your codes aren't working!

===> And about the URL of testing, will we continue to use that ` http://localhost:3000/api/auth/facebook/callback ` since the app will not run locally, but on heroku
.


### 2. Test coverage
You tried all of your best to make this task this far, but check again on the test coverage of your branch, it has decreased a lot. And there are a lot of uncovered lines of codes. Could you fix that? because even locally is behaving like that.
Screenshot from 2020-03-13 00-10-47

File: tests/socialLogin.test.js (hunk 4 above)
Comment (mifeille) [reply tests/socialLogin.test.js]: Please use descriptive names for those test routes, it will help everyone who reads your code to know that they are only there for testing.

Comment (Muhire-Josue) [reply tests/socialLogin.test.js]: Done. I think it should be easy to distinguish the test routes from others now.Thanks

**Review by since9629 — COMMENTED** [review]
@Muhire-Josue could you resolve conflicts this branch has with the base branch

File: .eslintrc.json [hunk 13]
```diff
@@ -13,16 +13,18 @@
...
-    "arrow-parens":0,
     "no-trailing-spaces":0,
     "no-useless-escape":0,
     "import/prefer-default-export": 0,
     "no-unused-vars": 0,
     "one-var-declaration-per-line": 0,
     "new-cap": 0,
     "consistent-return": 0,
+    "no-else-return":"off",
+    "arrow-parens":"off",
```
Comment (since9629) [inline .eslintrc.json]: why did you add these rules

File: .eslintrc.json [hunk 14]
```diff
@@ -13,16 +13,18 @@
...
     "no-param-reassign": 0,
     "comma-dangle": 0,
+    "no-underscore-dangle":"off",
```
Comment (since9629) [inline .eslintrc.json]: why did you add this rule

**Review by descholar-ceo — APPROVED** [review]

File: tests/config/passportConfig.test.js [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+/**
+   * @param {object} error
+   * @param {object} data
+   * @returns {object} response json object
+   */
+const callBackFunc = (error, data) => data;
+const user = { msg: 'hello' };
+
+describe('Passport Config', () => {
+  it('shoudl test Facebook passport', () => {
```
Comment (descholar-ceo) [inline tests/config/passportConfig.test.js:14]: maybe it is a typing error you wanted to type `should`

**Review by descholar-ceo — DISMISSED** [review]
I liked the work you did, it is great special congrats to you @Muhire-Josue,
I have two suggestions:
1. Could we use arrow functions consistently? is that possible? to make our codes consistent because you see for example in controllers and in services, there mixes of arrow functions and raw functions from ES5 syntax
1. While creating a user for the first login, on you are saving `NULL` on username, and this will be a big problem when it comes to displaying user's profile, as you see on the screenshot, even google even Facebook, all users have NULL as their username, remember the constraints of username (NOT NULL, UNIQUE), => So here I suggest that you use email as a default username as we talked about this before, and for users who don't have email (mostly from Facebook), you can generate a random username for them automatically but before saving it in the database, you check if that generated username is not already in DB.
Screenshot from 2020-03-16 18-13-48

You did a really great job, keep up and deal with those few remaining small tasks, then your feature will be great, thank you!

**Review by the22mastermind — COMMENTED** [review]
@Muhire-Josue Great job on this PR, I'm sure it wasn't easy. However, could you please provide more information on how your PR should be tested especially when creating and configuring apps on facebook and google. Also when signing in with facebook, the username field is left blank when it should be a required field. Could you generate a random username as @descholar-ceo suggested. Thanks. Keep up the good work.

**Review by descholar-ceo — APPROVED** [review]
Special congratulations on this long way you have made, this work was not easy however finally you made it, now the login with Facebook and Google work, and you can generate unique usernames for Facebook users because all of them don't have emails to set them as default usernames. That is great keep it up!

**Review by since9629 — APPROVED** [review]
Good job

File: tests/data/mockData.js [hunk 16]
```diff
@@ -94,5 +94,46 @@ export default {
...
+    username: 'Jdoe',
+    email: 'jane@doe.com',
+    address: 'Kigali',
+    password: 'helloworld3',
+    gender: 'Female',
+  },
+  usernameGeneratorUtil: {
+    firstName: 'Joe',
+    lastName: 'Doe',
+    // email: 'joe@doe.com',
```
Comment (ricahcyuzuzo) [inline tests/data/mockData.js]: Would remove this comment?

**Review by ricahcyuzuzo — APPROVED** [review]
Good Job Team mate!
